Improve OVAL test info

Honny1 opened the pull request:
This PR adds:

- OVAL state to OVAL test info
- check and check_existence attributes to OVAL test info

evgenyz commented:
There is a conflict. Can you please rebase the PR?

jan-cerny commented on openscap_report/report_generators/html_templates/js/oval_graph_generation_script.js:
```diff
@@ -532,6 +530,57 @@ function get_OVAL_test_info(test_info) {
+    div.appendChild(get_label("", test_info.test_id));
+    div.appendChild(get_label("pf-m-blue", test_info.comment));
+    if (test_info.check) {
+        div.appendChild(get_label("pf-m-cyan", `Check atribute: ${test_info.check}`));
```
It's nice that it shows now the value of the `check` and `check_existence` attributes. But I'm afraid that no user except from the content authors know what this means. It would be better to show also the meaning of that attribute.

It can, for example `check_existence="none_exist"`, say: This check requires that no such object can exist on the system.

You can find the possible values and their meaning in the OVAL specification. The report generate could translat attributus values to some helpful text.

jan-cerny commented on openscap_report/report_generators/html_templates/js/oval_graph_generation_script.js:
```diff
@@ -532,6 +530,57 @@ function get_OVAL_test_info(test_info) {
+    const div = DIV.cloneNode();
+    div.className = "pf-c-accordion__expanded-content-body";
+    div.appendChild(get_label("", test_info.test_id));
+    div.appendChild(get_label("pf-m-blue", test_info.comment));
```
The comment becames a tooltip  also for the check and check_existence rectangles. I don't think that it's useful to repeat it over and over. I think they could have a different tool tip, maybe related to the maining of the elements (see also the comment below).

jan-cerny reviewed (changes requested):
General feedback: it would be better to put some hints to the PR description for the reviewer regarding what the reviewer should focus on.

jan-cerny commented:
@Honny1 If you do some changes you need to summarize them in a comment in the PR so that the visitors know what changed.

Honny1 commented:
@jan-cerny Okay.

Changes:
* Removed shared static method for transition list of tuples to dictionary and create a dictionary in the first place,
* Created unit tests for shared methods,
* Reversed if condition in method `_get_ref_var`,
* Added tooltip info to `check`, `check_existence` and to id of objects and state. See the image below. 

![image](https://user-images.githubusercontent.com/26072444/219390704-7633d751-a3c5-4c1b-a51a-87d0a04fe779.png)

jan-cerny commented on openscap_report/scap_results_parser/parsers/oval_test_info_parser.py:
```diff
@@ -53,115 +60,30 @@ def _get_objects_by_id(self):
+                oval_object = self.objects_parser.get_object(object_id)
+            state_id = item.attrib.get("state_ref")
+            if state_id is not None:
+                oval_state = self.states_parser.get_state(state_id)
```
I don't know the details but I'm curious: Why is this a loop? In each iteration it overwrites the oval_object and oval_state. Wouldn't it be better to just pick the last item? And why is the `self.tests[test_id]` an iterable? Does that mean that there are more tests with the same ID?

Honny1 replied:
Yes you are right. This for loop is useless. The value contained in `self.tests` is a dictionary where the key is test_id and the value is an XML element. This loop is used to iterate through the children of the test XML element. I will simplify this code to make it more readable and replace the unnecessary loop with xPath.

jan-cerny commented on openscap_report/scap_results_parser/parsers/oval_object_parser.py:
```diff
@@ -0,0 +1,85 @@
+                out[id_in_dict] = self._get_ref_var(element, xml_collected_object)
+        return [out]
+
+    def _get_item(self, item_ref):
```
I find the name confusing, instead of "item" it returns some dictionary 

`out` is almost never a good name for anything

jan-cerny commented on openscap_report/scap_results_parser/parsers/oval_object_parser.py:
```diff
@@ -0,0 +1,85 @@
+                out[id_in_dict] = element.text
+            else:
+                out[id_in_dict] = self._get_ref_var(element, xml_collected_object)
+        return [out]
```
Why does this function return a list that contains a single item and that is a dictionary? That sounds suspicious and is a sign of a bad design. Please investigate this. If you really need this wrapping to a list it probably should be done one level of abstaraction above.


Also I suggest giving the variable a different name than `out`. Think what it actually contains and give it a descriptive name.

jan-cerny commented on openscap_report/scap_results_parser/parsers/oval_object_parser.py:
```diff
@@ -0,0 +1,85 @@
+        var_id = element.get('var_ref')
+        for item in xml_collected_object:
+            if var_id == item.get('variable_id'):
+                variable_value += item.text
```
More pythonic way of doing this would be to collect these in a list and after that merge the items in the list using `.join()` method.

jan-cerny commented on openscap_report/scap_results_parser/parsers/oval_state_parser.py:
```diff
@@ -0,0 +1,33 @@
+            operation = element.get("operation")
+            if operation is not None:
+                out["operation"] = operation
+        return [out]
```
this method has also similar understandability problems as the other similar methods - return value is a list that contains a single item which is a dict that has a generic name

Honny1 commented:
Changes:
* Improved tooltip text for `check` and `check_existence` attributes,
* renamed `out` variables,
* removed wrapping array in return,
* Used a more pythonic way to contact strings.

jan-cerny reviewed (approved):
I have generated a report for a fedora scan and I have checked the chekd and check_existence boxes in multiple rules and triggered their tooltips.

Great job!
